src/doc/trpl/closures.md

```diff
@@ -1,214 +1,513 @@
 % Closures
 
-So far, we've made lots of functions in Rust, but we've given them all names.
-Rust also allows us to create anonymous functions. Rust's anonymous
-functions are called *closures*. By themselves, closures aren't all that
-interesting, but when you combine them with functions that take closures as
-arguments, really powerful things are possible.
+Rust not only has named functions, but anonymous functions as well. Anonymous
+functions that have an associated environment are called 'closures', and Rust
+has a really great implementation of them, as we'll see.
 
-Let's make a closure:
+# Syntax
 
-```{rust}
-let add_one = |x| { 1 + x };
+Closures look like this:
 
-println!("The sum of 5 plus 1 is {}.", add_one(5));
+```rust
+let plus_one = |x| x + 1;
+
+assert_eq!(2, plus_one(1));
+```
+
+We create a binding, `plus_one`, and assign it to a closure. The closure's
+arguments go between the pipes (`|`), and the body is an expression, in this
+case, `x + 1`. Remember that `{ }` is an expression, so we can have multi-line
+closures too:
+
+```rust
+let plus_two = |x| {
+    let mut result = x;
+
+    result += 1;
+    result += 1;
+
+    result
+};
+
+assert_eq!(4, plus_two(2));
 ```
 
-We create a closure using the `|...| { ... }` syntax, and then we create a
-binding so we can use it later. Note that we call the function using the
-binding name and two parentheses, just like we would for a named function.
+You'll notice a few things about closures that are a bit different than regular
+functions defined with `fn`. The first of which is that we did not need to
+annotate the types of arguments the closure takes or the values it returns. We
+can:
 
-Let's compare syntax. The two are pretty close:
+```rust
+let plus_one = |x: i32| -> i32 { x + 1 };
 
-```{rust}
-let add_one = |x: i32| -> i32 { 1 + x };
-fn  add_one   (x: i32) -> i32 { 1 + x }
+assert_eq!(2, plus_one(1));
 ```
 
-As you may have noticed, closures infer their argument and return types, so you
-don't need to declare one. This is different from named functions, which
-default to returning unit (`()`).
+But we don't have to. Why is this? Basically, it was chosen for ergonomic reasons.
+While specifying the full type for named functions is helpful, closures aren't going
+to, for example, appear in Rustdoc output somewhere. They don't cause the kinds of
+error-at-a-distance that inferring named function types can. In practice, it
+works well.
 
-There's one big difference between a closure and named functions, and it's in
-the name: a closure "closes over its environment." What does that mean? It means
-this:
+The second is that the syntax is similar, but a bit different. I've added spaces
```

Comment:
This may be really really helpful depending on your background. Originally, it took me a [lot](https://github.com/rust-lang/rust/pull/13676#commitcomment-6074211) of time to discover they can be likened in that way. I found that syntax really really helpful. For example, two questions I had before the syntax was aligned that way from that link:

> * Why is the absolute value of `arg` multiplied by `println!(...)` and why does it result in a closure?
* Why is the times sign (`*`) missing from `arg` times `println!(...)`?
>
> `let closure = |arg| println!("captured_var={}, arg={}", captured_var, arg);`

The lack of return type was especially odd and confusing. By the way, I think I was probably familiar with anonymous function but how it was being utilized/explained was completely foreign to me.